src/track/serato/tags.cpp

```diff
@@ -0,0 +1,82 @@
+#include "track/serato/tags.h"
+
+namespace mixxx {
+
+RgbColor::optional_t SeratoTags::colorFromStoredTrackColor(RgbColor color) {
+    // Serato stores Track colors differently from how they are displayed in
+    // the library column. Instead of the color from the library view, the
+    // value from the color picker is stored instead (which is different).
+    // To make sure that the track looks the same in both Mixxx' and Serato's
+    // libraries, we need to convert between the two values.
+    //
+    // See this for details:
+    // https://github.com/Holzhaus/serato-tags/blob/master/docs/colors.md#track-colors
+
+    quint32 value = static_cast<quint32>(color);
+
+    if (color == 0xFFFFFF) {
+        return RgbColor::nullopt();
+    }
+
+    if (color == 0x999999) {
+        return RgbColor::optional(0x090909);
+    }
+
+    if (color == 0x000000) {
+        return RgbColor::optional(0x333333);
+    }
+
+    value = (value < 0x666666) ? value + 0x99999A : value - 0x666666;
+    return RgbColor::optional(value);
+}
+
+RgbColor SeratoTags::colorToStoredTrackColor(RgbColor::optional_t color) {
+    if (!color) {
+        return RgbColor(0xFFFFFF);
+    }
+
+    quint32 value = static_cast<quint32>(*color);
+
+    if (value == 0x090909) {
+        return RgbColor(0x999999);
+    }
+
+    if (value == 0x333333) {
+        return RgbColor(0x000000);
+    }
+
+    // Special case: 0x999999 and 0x99999a are not representable as Serato
+    // track color We'll just modify them a little, so that the look the
+    // same in Serato.
+    if (value == 0x999999) {
+        return RgbColor(0x999998);
+    }
+
+    if (value == 0x99999a) {
+        return RgbColor(0x99999b);
+    }
+
+    value = (value < 0x99999A) ? value + 0x666666 : value - 0x99999A;
+    return RgbColor(value);
+}
+
+RgbColor::optional_t SeratoTags::getTrackColor() const {
+    RgbColor::optional_t color = m_seratoMarkers.getTrackColor();
+
+    if (!color) {
```

Comment:
1st step: Obtain color from one of the markers
2nd step: Transform and return color

The mutable `color` variable is highly confusing. At first sight the control flow suggests unreachable code. Please improve this.

Reply:
This function should be split. Both getTrackColor() functions should already return decoded and valid color values. Stuffing some cryptic integer into the well-defined RgbColor type is highly unsafe and may cause all kinds of bugs.

Reply:
To clarify: Both `getTrackColor()` functions *do* return valid color values. But for some reason Serato saves the color that is shown in the color picker, not the color displayed in the library column. See here: https://github.com/Holzhaus/serato-tags/blob/master/docs/colors.md#track-colors

That said, I can still move the conversion into the individual `getTrackColor()` functions if you want. Just tell me.

Reply:
Aweome reverse engineering! I didn't know this and confused it with their cryptic tag encoding. 

WTH did they had in mind when they implemented this??

The term "track color" might be misleading and is ambiguous. How about to invert the naming and use "display color" instead? Could this help to not fall into this trap?

Reply:
Which one of the two colors is the reference and which should only be used temporary?

Reply:
Serato users only see the *stored* color when assigning track colors - most of the time they'll just see the *displayed* color in the library column. IMHO we should show the latter.